internal/controller/apiextensions/managed/reconciler.go

```diff
@@ -0,0 +1,219 @@
+/*
+Copyright 2025 The Crossplane Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package managed manages the lifecycle of MRD controllers.
+package managed
+
+import (
+	"context"
+	"time"
+
+	extv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/api/equality"
+	kerrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/crossplane/crossplane-runtime/pkg/conditions"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/crossplane/apis/apiextensions/v2alpha1"
+	"github.com/crossplane/crossplane/internal/controller/apiextensions/managed/resources"
+	"github.com/crossplane/crossplane/internal/xcrd"
+)
+
+const (
+	timeout = 2 * time.Minute
+
+	errGetMRD          = "cannot get ManagedResourceDefinition"
+	errUpdateStatus    = "cannot update status of ManagedResourceDefinition"
+	errGetCRD          = "cannot get CustomResourceDefinition"
+	errMergeCRD        = "cannot merge CustomResourceDefinition"
+	errReconcileCRD    = "cannot reconcile CustomResourceDefinition"
+	errCreateCRD       = "cannot create CustomResourceDefinition"
+	errUpdateCRD       = "cannot update CustomResourceDefinition"
+	errInvalidResource = "invalid resource configuration"
```

Comment:
I'd prefer to avoid these err constants and inline the errors. We made a decision to move away from this pattern and have been slowly doing so as we introduce new controllers (context in contributing/) iirc.

(Now that we have the magic of AI I've been meaning to ask an agent to go inline all the error constants.)